### ISSUE TITLE
Entities vs EXTERNAL vs PROCEDURE 

### ISSUE THREAD

**Issue by schauveau** [flang-compiler/f18#72]
The purpose of this ticket is to discuss the implementation of the EXTERNAL attribute and the new PROCEDURE( )  declaration statement. 

First a bit of history: In early Fortran,  calling an external function was easy. You would simply declare its non-array return type and use it as a function:
 
    integer foo
    print *, foo(42)

The entity foo was known to be a function by its usage.
Of course, using a declared entity as a function and as variable is not possible:
    
    integer foo, bar
    print *, bar        ! bar is now a variable
    print *, foo(42)  ! legal: foo is now external
    print *, bar(42) ! illegal: bar was previously used as a variable 

So the first consequence of this behavior is that we need to track the use of each entity (local declaration and dummy arguments) that is susceptible to gain the EXTERNAL attribute. I propose to add a new attribute VARIABLE that should initially be false and that that will be set to true once the entity is used as a variable.

At that point, an EXTERNAL attribute was introduced to allow external procedures to be passed as arguments.

    integer foo, val
    external foo 
    call bar(foo,val) ! foo is an external while val is a variable 
    ...
    subroutine bar(fun)
       integer fun, val
       print *, fun(val)
    end

We see here that the dummy argument 'fun' gained the EXTERNAL attribute. 

In modern Fortran a new syntax was introduced to declare PROCEDURE entities. 
For instance, the following is strictly equivalent to INTEGER, EXTERNAL :  
 
    procedure(integer) :: foo 

* PROCEDURE() can also take a procedure name as argument to specify a full prototype instead of a return type. 
* a PROCEDURE() entity can have the POINTER attribute. 

The current implementation is still incomplete but 

* In ResolveNamesVisitor::Post(const parser::SubroutineStmt &stmt), all dummy arguments symbols  are immediately given EntityDetails 
* In Post(const parser::ProcedureDesignator &x), symbols without 'details' are given SubprogramDetails. 

Also, when given explicitly, the EXTERNAL attribute appears to be treated like any other attributes.

I am proposing something slightly different: 

1) add a new Procedure category to DeclTypeSpec with two variants: one that provides only a return type and one that provides a Symbol with SubprogramDetails 

2) An entity declared explicitly with PROCEDURE(xxxx) shall be associated to a Symbol with EntityDetails and a DeclTypeSpec of category Procedure(xxxx).

3) when an entity is given the EXTERNAL attribute (either explicitly or by usage), its symbol should be given EntityDetails (if this is not already the case) and its type should be changed from 'oldtype' to 'Procedure(oldtype)' 

4) For each subroutine or function construct, the associated Symbol 'name' should be given  SubprogramDetails. Querying the type of that symbol shall return a DeclTypeSpec of  category Procedure(name).

To summarize, I propose to encode 'external' and procedure() in the symbol type instead of using an attribute.
